duckduckgo/Android
# Use providers in the VM factories to respect the dep di semantics

Task/Issue URL: https://app.asana.com/0/30173902528854/1200098175084894/f
Tech Design URL: 
CC: 

**Description**:
https://github.com/duckduckgo/Android/pull/1123 introduced a bug that was detected by our pixels.

* That pull request delegated the creation of the view models (VMs) to its own factory
* The VM factories are `singleton` -- as they should be
* The VM factories create the VMs and inject their dependencies
* Because the VM factories are `singleton` that effectively makes all VM dependencies `singleton`. Regardless of whether they are actually declare as such or not
* This is not desired, VM factories thouls be singleton, but it should respect the di semantics of its deps

This PR makes all the factories to provide `providers` of the dependencies, rather than the actual instance. So that they respect how the instances declare themselves when injected, ie. `singleton` or multiple instances.


In particular, the effects can be seen with the `fw_` pixels. They are fired multiple times (one per fragment) when they should fire only once.

**Steps to test this PR**:
1. App compiles
2. install/upgrade the app from this branch
3. use the SERP to perform searches
4. form the SERP, open several tabs
5. add/remove bookmarks
6. add/remove fireproof sites
7. navigate to settings screen
8. use fire button
10. clear app data and go through onboarding UX
11. for all above, verify app works as expected


1. install/upgrade from this branch
2. open several tabs and ensure they are active
3. ensure `Ask When Signing In` is enabled inside `settings -> fireproof sites`
4. open a new tab and log into facebook.com (or any other login you may want)
5. verify the dialog to fireproof the site appears
6. navigate to any other of the active tabs
7. verify no dialog appears
8. verify `m_fw_xx` are only shown once



---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
